tsl/src/bgw_policy/job.c

```diff
@@ -513,24 +513,60 @@ policy_invoke_recompress_chunk(Chunk *chunk)
 bool
 policy_compression_execute(int32 job_id, Jsonb *config)
 {
-	int32 chunkid;
+	List *chunkid_lst;
+	ListCell *lc;
 	const Dimension *dim;
 	PolicyCompressionData policy_data;
+	bool distributed, used_portalcxt = false;
+	MemoryContext saved_cxt, multitxn_cxt;
 
 	policy_compression_read_and_validate_config(config, &policy_data);
 	dim = hyperspace_get_open_dimension(policy_data.hypertable->space, 0);
-	chunkid = get_chunk_to_compress(dim, config);
+	distributed = hypertable_is_distributed(policy_data.hypertable);
+	/* we want the chunk id list to survive across transactions. So alloc in
+	 * a different context
+	 */
+	if (PortalContext)
+	{
+		/*if we have a portal context use that - it will get freed automatically*/
+		multitxn_cxt = PortalContext;
+		used_portalcxt = true;
+	}
+	else
+	{
+		/* background worker job does not go via usual CALL path, so we do
+		 * not have a PortalContext */
+		multitxn_cxt =
+			AllocSetContextCreate(TopMemoryContext, "CompressionJobCxt", ALLOCSET_DEFAULT_SIZES);
```

Comment:
hmm should we create this below CurrentMemoryContext instead?

Reply:
The current memory context is TopMmeoryContext here. However, I think it is safer to explicitly reference TopMemoryContext here.

Reply:
I don't think it is safe to assume the CurrentMemoryContext has a lifetime beyond the current transaction. TopMemoryContext is guaranteed to exist across transactions